### DISCUSSION TITLE
QST: read_file unexpectedly only reads GeoPackage headers and not data or geometry

### DISCUSSION THREAD

**Discussion by EwoutH** [geopandas/geopandas#3605] [Q&A]
- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of geopandas.
- [ ] (optional) I have confirmed this bug exists on the master branch of geopandas.

---

#### Problem (with code)

I tried reading a GeoPackage from CSB with [Dutch Neighbourhoods](https://www.cbs.nl/nl-nl/dossier/nederland-regionaal/geografische-data/wijk-en-buurtkaart-2021), which only loaded the headers and not the data and geometries.
```python
neighbourhoods = gpd.read_file("WijkBuurtkaart_2021_v1\WijkBuurtkaart_2021_v1.gpkg", layer='cbs_wijken_2021')
```
Which resulted in an empty geodataframe with only headers:
![Screenshot_725](https://user-images.githubusercontent.com/15776622/138672146-1866e240-06e3-4c4b-a199-0118dc2c6b74.png)

#### Expected Output

A full geodataframe containing data and geometries.

I tested reading the GeoPackage in ArcGIS Pro, and there it loaded correctly.

![Screenshot_724](https://user-images.githubusercontent.com/15776622/138671558-00c31bd0-b9aa-46b9-b3b8-ca07c026ce3b.png)

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:15:42) [MSC v.1916 64 bit (AMD64)]
executable : C:\Users\Ewout\anaconda3\envs\py39\python.exe
machine    : Windows-10-10.0.19043-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.3.2
GDAL data dir: None
PROJ       : 8.1.1
PROJ data dir: C:\Users\Ewout\anaconda3\envs\py39\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.2
pandas     : 1.3.4
fiona      : 1.8.20
numpy      : 1.20.3
shapely    : 1.7.1
rtree      : 0.9.7
pyproj     : 3.2.1
matplotlib : 3.4.3
mapclassify: 2.4.3
geopy      : 2.2.0
psycopg2   : None
geoalchemy2: None
pyarrow    : None
pygeos     : 0.10.2

</details>

CC @MarcStreng.

**Comment by jorisvandenbossche** [comment]
@EwoutH I downloaded the file and I can reproduce the issue you see. 

There seems to be _something_ about the file that makes GDAL/OGR (the library which is being used to read the data, through the fiona python package) incorrectly report the feature count. 

When using GDAL's command line interface to get a summary of the file, I get:

```
$ ogrinfo -so WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg cbs_wijken_2021
INFO: Open of `WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg'
      using driver `GPKG' successful.

Layer name: cbs_wijken_2021
Geometry: Multi Polygon
Feature Count: 0
Extent: (10425.154297, 306846.187500) - (278026.093750, 621876.312500)
Layer SRS WKT:
PROJCRS["Amersfoort / RD New",
...
```

<details>
<summary>Full output:</summary>

```
$ ogrinfo -so WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg cbs_wijken_2021
INFO: Open of `WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg'
      using driver `GPKG' successful.

Layer name: cbs_wijken_2021
Geometry: Multi Polygon
Feature Count: 0
Extent: (10425.154297, 306846.187500) - (278026.093750, 621876.312500)
Layer SRS WKT:
PROJCRS["Amersfoort / RD New",
    BASEGEOGCRS["Amersfoort",
        DATUM["Amersfoort",
            ELLIPSOID["Bessel 1841",6377397.155,299.1528128,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4289]],
    CONVERSION["RD New",
        METHOD["Oblique Stereographic",
            ID["EPSG",9809]],
        PARAMETER["Latitude of natural origin",52.1561605555556,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",5.38763888888889,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9999079,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",155000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",463000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["easting (X)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["northing (Y)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["Engineering survey, topographic mapping."],
        AREA["Netherlands - onshore, including Waddenzee, Dutch Wadden Islands and 12-mile offshore coastal zone."],
        BBOX[50.75,3.2,53.7,7.22]],
    ID["EPSG",28992]]
Data axis to CRS axis mapping: 1,2
FID Column = fid
Geometry Column = geom
wijkcode: String (8.0)
wijknaam: String (60.0)
gemeentecode: String (6.0)
gemeentenaam: String (60.0)
indelingswijziging_wijken_en_buurten: Integer (0.0)
water: String (4.0)
omgevingsadressendichtheid: Integer (0.0)
stedelijkheid_adressen_per_km2: Integer (0.0)
bevolkingsdichtheid_inwoners_per_km2: Integer (0.0)
aantal_inwoners: Integer (0.0)
mannen: Integer (0.0)
vrouwen: Integer (0.0)
percentage_personen_0_tot_15_jaar: Integer (0.0)
percentage_personen_15_tot_25_jaar: Integer (0.0)
percentage_personen_25_tot_45_jaar: Integer (0.0)
percentage_personen_45_tot_65_jaar: Integer (0.0)
percentage_personen_65_jaar_en_ouder: Integer (0.0)
percentage_ongehuwd: Integer (0.0)
percentage_gehuwd: Integer (0.0)
percentage_gescheid: Integer (0.0)
percentage_verweduwd: Integer (0.0)
aantal_huishoudens: Integer (0.0)
percentage_eenpersoonshuishoudens: Integer (0.0)
percentage_huishoudens_zonder_kinderen: Integer (0.0)
percentage_huishoudens_met_kinderen: Integer (0.0)
gemiddelde_huishoudsgrootte: Real (0.0)
percentage_westerse_migratieachtergrond: Integer (0.0)
percentage_niet_westerse_migratieachtergrond: Integer (0.0)
percentage_uit_marokko: Integer (0.0)
percentage_uit_nederlandse_antillen_en_aruba: Integer (0.0)
percentage_uit_suriname: Integer (0.0)
percentage_uit_turkije: Integer (0.0)
percentage_overige_nietwestersemigratieachtergrond: Integer (0.0)
oppervlakte_totaal_in_ha: Integer (0.0)
oppervlakte_land_in_ha: Integer (0.0)
oppervlakte_water_in_ha: Integer (0.0)
jaarstatcode: String (15.0)
jaar: Integer (0.0)
```

</details>

So there is a `Feature Count: 0` in there. 

However, when actually asking the full data (and not a summary overview like the above), it does return features. For example only asking the first feature (truncating the output a bit):

```
$ ogrinfo -fid 1 WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg cbs_wijken_2021
INFO: Open of `WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg'
      using driver `GPKG' successful.

Layer name: cbs_wijken_2021
Geometry: Multi Polygon
Feature Count: 0
Extent: (10425.154297, 306846.187500) - (278026.093750, 621876.312500)
Layer SRS WKT:
PROJCRS["Amersfoort / RD New",
...
Data axis to CRS axis mapping: 1,2
FID Column = fid
Geometry Column = geom
wijkcode: String (8.0)
wijknaam: String (60.0)
...
jaar: Integer (0.0)
OGRFeature(cbs_wijken_2021):1
  wijkcode (String) = WK003499
  wijknaam (String) =  
...
  jaar (Integer) = 2021
  MULTIPOLYGON (((150213.998 479503.726,150087.298999999 479382.379000001,150000.420000002
...
```

So while the data is there, the feature count is giving (for some reason I don't know) zero. And I that's the reason that geopandas / fiona is not reading in any data, as it will basically do `for i in range(feature_count): ...` to process each of the features in the dataset.

**Comment by brendan-ward** [comment]
I believe that it is reporting 0 features for each data layer within the GPKG because the `gpkg_ogr_contents` table in the SQLite file stores it that way (how that got set is unclear):
![image](https://user-images.githubusercontent.com/3375604/138766819-0bc90fc4-a9e7-4c6f-b829-383757a5b6a2.png)

Manually changing contents of that table to the correct value corrects that issue.  If you use `ogr2ogr2` to convert this into a new GPKG, it sets that value correctly.

```
ogr2ogr -f GPKG /tmp/test.gpkg WijkBuurtkaart_2021_v1.gpkg cbs_wijken_2021
```

**Comment by jorisvandenbossche** [comment]
Searching the GDAL repo for similar issues, I found https://github.com/OSGeo/gdal/issues/4256, which led me to take a look at the `gpkg_ogr_contents` table in the GeoPackage file. Manually inspecting that table by using sqlite3 in Python to read from the GeoPackage file:

```python
>>> import sqlite3
>>> con = sqlite3.connect("WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg")
>>> con.execute("SELECT * FROM gpkg_ogr_contents;").fetchall()
[('cbs_buurten_2021', 0), ('cbs_wijken_2021', 0), ('gemeenten2021', 0)]
```

So the problem is that this table is not up to date with the actual contents of the file. 

I _suppose_ this is an issue with how the file was created, and you should maybe report this to the Dutch administration that created the dataset.

**Comment by jorisvandenbossche** [comment]
A small workaround in case you know the exact number of features:

```python
>>>con.execute("UPDATE gpkg_ogr_contents SET feature_count = 5; ")
<sqlite3.Cursor at 0x7f9bf386a500>
>>> con.commit()
>>> geopandas.read_file("WijkBuurtkaart_2021_v1/WijkBuurtkaart_2021_v1.gpkg")
    buurtcode buurtnaam  wijkcode gemeentecode     gemeentenaam  ...  oppervlakte_land_in_ha oppervlakte_water_in_ha    jaarstatcode  jaar                                           geometry
0  BU00349997            WK003499       GM0034           Almere  ...               -99999999               -99999999  2021BU00349997  2021  MULTIPOLYGON (((150213.998 479503.726, 150087....
1  BU00509997            WK005099       GM0050         Zeewolde  ...               -99999999               -99999999  2021BU00509997  2021  MULTIPOLYGON (((170588.413 486792.192, 170570....
2  BU00609998            WK006099       GM0060          Ameland  ...               -99999999               -99999999  2021BU00609998  2021  MULTIPOLYGON (((196000.000 610000.000, 196000....
3  BU00729998            WK007299       GM0072        Harlingen  ...               -99999999               -99999999  2021BU00729998  2021  MULTIPOLYGON (((158392.775 580357.500, 158387....
4  BU00889998            WK008899       GM0088  Schiermonnikoog  ...               -99999999               -99999999  2021BU00889998  2021  MULTIPOLYGON (((219000.000 616567.418, 219000....

**Comment by jorisvandenbossche** [comment]
@brendan-ward missed your comment while researching, and indeed we came to the same conclusion. But your workaround of using `ogr2ogr` is much better (more robust) than manually fiddling with the feature count!

**Comment by EwoutH** [comment]
Thank you both for looking into this! I have emailed the CBS, who published the dataset, and will update you when they respond.

Should handeling GeoPackages with a wrong Feature Count be something GDAL and/or OGR should be capable of, and if so, should a bug be also opened there?

**Comment by brendan-ward** [comment]
GDAL / OGR are likely doing expected behavior here; if there is a table `gpkg_ogr_contents` present, this table is expected to have valid counts, so it seems reasonable to use them rather than independently verify them.

Instead, it is likely an issue with how the dataset was created in the first place; if it was a GDAL / OGR bug on the _creation_ side, then that indeed would be an issue to file with the GDAL project.

**Comment by cmosig** [comment]
I experience the same behavior with a different dataset. I am also able to fix the dataset with `ogr2ogr`. Some kind of warning in geopandas that the feature count is incorrect may be helpful to unassuming users.

**Comment by theroggy** [comment]
For what it is worth... there has been a fix in GDAL >= 3.10.1 that is somewhat related, but at first sight it doesn't seem to be the exact same as what might be happening here.

In some cases, GDAL created GeoPackages with `featurecount = NULL` instead of `featurecount = 0`, which can e.g. lead to performance issues. This was fixed here: https://github.com/OSGeo/gdal/pull/11275

In this discussion it seems that GeoPackages were created with `featurecount = 0`, which is a different case. However, based on the information above, we don't even know if the file was created using GDAL... so I don't think there is anything that can be solved...
